---
output: github_document

---

```{r setup, include=FALSE, cache=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.path = "LifespanAnalysis/Lifespan_v1.0.0-",
                      out.width = "100%",
                      cache = TRUE,
                      cache.lazy = FALSE)
# for tibbles...
options(pillar.neg=F, # do no print neg number in red
        pillar.subtle=F, # turn off highliting of significant digits
        tibble.width = 130) # default=95, increase it to make it readable
```


# Lifespan Analysis

In this example, we will use the data from Timmers et al to apply our Bayesian GWAS approach to study lifespan.    
Here, we assume that the `bGWAS` package is already installed, that the Z-matrix files have already been downloaded and stored in "~/ZMatrices". If that is not the case, please follow the steps described [here](../README.md). 

```{r download-data, echo=TRUE, message=FALSE, results='hide', cache=F}
library(bGWAS) # bGWAS github version:

# Download data to working directory (~460 MB) if not already here
print(getwd())
if(!file.exists("lifegen_phase2_bothpl_alldr_2017_09_18.tsv.gz")) download.file(url = "https://datashare.is.ed.ac.uk/bitstream/handle/10283/3209/lifegen_phase2_bothpl_alldr_2017_09_18.tsv.gz?sequence=1&isAllowed=y", destfile = "lifegen_phase2_bothpl_alldr_2017_09_18.tsv.gz")

Lifespan_Timmers2019 = "lifegen_phase2_bothpl_alldr_2017_09_18.tsv.gz"
```

Now that we have the data in our working directory, we can launch the analysis (with default parameters):

```{r bGWAS, echo=TRUE}
print(getwd())
Lifespan_bGWAS = bGWAS(name = "Lifespan_Timmers2019",
                       GWAS = Lifespan_Timmers2019)
```



We can look at the results more in details.

## Prior GWASs used

```{r results1, dpi=300}
print(getwd())
coefficients_plot_bGWAS(Lifespan_bGWAS)
```

7 prior GWASs are used to create the prior, the multivariate causal effect estimates are consistent with what we would expect. On this figure, the multivariate causal effect estimate and the 95\% interval from the multivariate MR model using all chromosomes (black dot and bars) as well as the per-chromosome estimates (grey bars) are represented for each prior GWASs. Coronary Artery Disease (CAD) has the strongest negative effect on lifespan. High Diastolic Blood Pressure (DBP) and Body Mass Index (BMI) also decreases lifespan. We can also see that education, in this case the number of years of schooling, has a positive effect on lifespan.

Overall, the squared correlation between prior and observed effects is about `r round(get_RSquared_bGWAS(Lifespan_bGWAS, "all"), 3)` and goes up to `r round(get_RSquared_bGWAS(Lifespan_bGWAS, "moderate"), 3)` when we consider only SNPs having at least a moderate effect on lifespan (observed p-value < 0.001).

## Results - BF


With this approach, we identified `r nrow(extract_results_bGWAS(Lifespan_bGWAS))`:

```{r results2, echo=TRUE}
# all hits
knitr::kable(extract_results_bGWAS(Lifespan_bGWAS) %>% mutate(BF_p=as.character(format(BF_p, scientific=T))))

# new hits (compared to conventional GWAS)
extract_results_bGWAS(Lifespan_bGWAS) %>%
  mutate(obs_p = 2*pnorm(-abs(z_obs))) %>%
  filter(obs_p>5e-8) -> NEW
knitr::kable(NEW %>% mutate(BF_p=as.character(format(BF_p, scientific=T))))
```

`r nrow(NEW)` of them are missed by the conventional GWAS (using same p-value threshold of 5e-8 to call significance).


```{r results3, dpi=300}
manhattan_plot_bGWAS(Lifespan_bGWAS)
```

```{r results4, dpi=300}
heatmap_bGWAS(Lifespan_bGWAS)
```


## Results - Direct Effects


We can use direct effects to identify SNPs significantly acting on lifespan independently from the prior GWASs used to create the prior:

```{r results5, echo=TRUE}
knitr::kable(extract_results_bGWAS(Lifespan_bGWAS, "direct")  %>% mutate(p_direct=as.character(format(p_direct, scientific=T))))
```

Among these hits...: Not APOE (or CHRNA, but smoking not used so makes sense?) ... needs to be looked at a bit more!    



\newpage
\Large \textbf{References}
\scriptsize
